crypt.5

```diff
@@ -152,11 +152,12 @@ and does not show the division into prefix, options, salt, and hash.
 .ti -4
 .B yescrypt
 .br
-yescrypt is a scalable password hashing scheme designed by Solar
-Designer, which is based on Colin Percival's scrypt.
-.hash "$y$" "\e$y\e$[./A-Za-z0-9]+\e$[./A-Za-z0-9]*\e$[./A-Za-z0-9]{43}" unlimited 8 "" 256 "up to 512" "0 (default), 1 to 11 (logarithmic)"
+yescrypt is a scalable password hashing scheme designed by
+Solar Designer, which is based on Colin Percival's scrypt.
+.hash "$y$" "\e$y\e$[./A-Za-z0-9]+\e$[./A-Za-z0-9]{25,…}\e$[./A-Za-z0-9]{43}" unlimited 8 256 256 "128 to (theoretically) 17895696" "1 to 11 (logarithmic)"
```

Comment:
This definitely should be a valid regexp, and besides the changes look wrong to me, while the original by @vt-alt looked correct to me.

Reply:
FWIW, I think I'm the original author of the `.hash` macro, first created for Owl in 2000, but I'm not yet familiar with how it's changed in libxcrypt since then - the word "effectively" isn't in my version, and instead my version doesn't confuse limit on effective input key size and hash size, which are not the same - maybe this is a bug introduced in libxcrypt that needs to be fixed (someone forgot about salts when thinking of the effect of key size on effective hash size?) Proper reporting for "Traditional DES-based" on Owl is.:

```
       prefix "" (empty string);
              a string matching ^[./0-9A-Za-z]{2} (see regex(7))

       Encoding syntax
              [./0-9A-Za-z]{13}

       Maximum password length
              8 (uses 7-bit characters)

       Effective key size
              up to 56 bits

       Hash size
              64 bits

       Salt size
              12 bits

       Iteration count
              25
```

yescrypt's salt size is limited to 512 bits max through `unsigned char saltbin[64]` in `yescrypt_r()`, which is chosen this way in part because yescrypt's builtin salt+hash encryption feature (which we don't expose here) also has this limit for more fundamental reason (it's max block size of the variable-block-size cipher used there, and there's no extra code to implement a block cipher mode in there, so we always encrypt/decrypt just one block).

I'd rather not introduce a minimum salt size in order to make the crypt(3)-encoding interface to yescrypt almost as capable as the KDF interface (where allowing empty salts is also needed for compatibility with scrypt and makes sense for some use cases), without artificial limitations. (The upper limit of 512 bits isn't entirely artificial - there would have to be some since the salt is encoded/decoded, and sometimes even encrypted/decrypted.)

Reply:
For comparison, here is what we get for DES with the current `.hash` macro in libxcrypt's crypt.5:

```
       prefix "" (empty string)

       Encoded passphrase format
              [./0-9A-Za-z]{13}

       Maximum password length
              8 characters (ignores 8th bit)

       Hash size
              64 bits (effectively 56)

       Salt size
              12 bits

       CPU time cost parameter
              25
```

> my version doesn't confuse limit on effective input key size and hash size, which are not the same - maybe this is a bug introduced in libxcrypt that needs to be fixed (someone forgot about salts when thinking of the effect of key size on effective hash size?)

I don't understand this.  The "hash size" is the size of the output of the one-way function, which is independent of the size of the salt, isn't it?  Anyway, let's split this off to a separate discussion: whether or not the hash size / key size terminology currently used in `crypt.5` is ideal or even correct, I hope you'll agree that the table entry for yescrypt should not read 

```
       Hash size
              256 bits (effectively )
```

and that's the only thing I was concerned with here.  Would you please read through `man -l crypt.5` using our crypt.5, and file a fresh issue describing any concerns you have?

> yescrypt's salt size is limited to 512 bits max through unsigned char saltbin[64] in yescrypt_r(), which is chosen this way in part because yescrypt's builtin salt+hash encryption feature (which we don't expose here) also has this limit for more fundamental reason

Aha.  Thanks.  I think that means the correct regex is

    \$y\$[./A-Za-z0-9]+\$[./A-Za-z0-9]{,86}\$[./A-Za-z0-9]{43}

and the salt size should remain "up to 512 bits" as it originally was.  Why doesn't yescrypt_encode_params_r enforce that limit, though?

> I'd rather not introduce a minimum salt size in order to make the crypt(3)-encoding interface to yescrypt almost as capable as the KDF interface

Fair enough.

Reply:
Yes, I agree the "256 bits (effectively )" display is obviously wrong, and the reporting of "64 bits (effectively 56)" for descrypt, which I think is wrong, should be discussed separately. Your suggested regexp for yescrypt looks correct to me, although I'd consistently list 0-9 before A-Z in these regexp's as that's the order used in the encoding (except for bcrypt's, where what you wrote here for yescrypt would be "more correct"). Good point about possibly enforcing the salt size limit in `yescrypt_encode_params_r()`, I'll consider making this change.